asmrun/afl.c

```diff
@@ -0,0 +1,100 @@
+/* Runtime support for afl-fuzz */
+
+#ifdef _WIN32
+
+void caml_maybe_setup_afl () {}
+
+#else
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/shm.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include "caml/misc.h"
+
+/* Values used by the instrumentation logic (see cmmgen.ml) */
+static unsigned char afl_area_initial[1 << 16];
+unsigned char* caml_afl_area_ptr = afl_area_initial;
+uintnat caml_afl_prev_loc;
+
+/* File descriptors used to synchronise with afl-fuzz */
+#define FORKSRV_FD_READ 198
+#define FORKSRV_FD_WRITE 199
+
+static void afl_write(uint32_t msg)
+{
+  if (write(FORKSRV_FD_WRITE, &msg, 4) != 4)
+    caml_fatal_error("writing to afl-fuzz");
+}
+
+static uint32_t afl_read()
+{
+  uint32_t msg;
+  if (read(FORKSRV_FD_READ, &msg, 4) != 4)
+    caml_fatal_error("reading from afl-fuzz");
+  return msg;
+}
+
+void caml_maybe_setup_afl()
+{
```

Comment:
There needs to be a check here that the binary was instrumented for AFL.  Otherwise executing
```sh
__AFL_SHM_ID=10a <any native-code OCaml program>
```
would cause the program to crash.